### PR TITLE
74 - Refactor the submission repository to be xpub schema specific for submission root entries

### LINKED ISSUE #74: Refactor submission repo to do xpub schema mapping and only take / return a DTO object
Currently we are mapping data from / to the xpub schema format outside of the repo. This should be the repos responsibility as no other part of the app code should need to know about the xpub schema layout. W should also define a standard DTO interface to be passed to the repository and returned from the repository. This will need a mapper class in the models layer for converting to / from object -> DTO and vise-versa.

### PR DESCRIPTION
**Pull request by Cooryd**
Closes #74

### CODE REVIEW

File: src/submission/infrastructure/types.ts [hunk 1]
```diff
@@ -0,0 +1,20 @@
+import { uuidType } from 'typesafe-uuid';
+
+export class SubmissionId extends uuidType<'SubmissionId'>() {}
+
+export interface SubmissionRepository {
+    findAll(): Promise<SubmissionDTO[]>;
+    findById(id: SubmissionId): Promise<SubmissionDTO | null>;
+    save(sub: SubmissionDTO): Promise<SubmissionDTO | null>;
```
Comment (diversemix) [inline src/submission/infrastructure/types.ts]: should `save` be replaced with `create` and `update` (following the CRUD pattern)

File: src/submission/resolvers/submission.ts [hunk 2]
```diff
@@ -8,7 +8,7 @@ const resolvers = (submissionService: SubmissionService): IResolvers => ({
             return await submissionService.findAll();
         },
         async getSubmission(id: string): Promise<DtoViewSubmission | null> {
-            return await submissionService.findOne(SubmissionId.fromUuid(id));
+            return await submissionService.getSubmission(SubmissionId.fromUuid(id));
```
Comment (diversemix) [inline src/submission/resolvers/submission.ts]: "Submission" is redundant on `getSubmission` favour using `submissionService.get()`

File: src/submission/resolvers/submission.ts [hunk 3]
```diff
@@ -21,7 +21,7 @@ const resolvers = (submissionService: SubmissionService): IResolvers => ({
         },
 
         async deleteSubmission(_, { id }: { id: SubmissionId }): Promise<SubmissionId> {
-            await submissionService.delete(id);
+            await submissionService.deleteSubmission(id);
```
Comment (diversemix) [inline src/submission/resolvers/submission.ts]: ditto above comment for `delete`

File: src/submission/repositories/xpub-submission-root.ts [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+    }
+
+    public async findById(id: SubmissionId): Promise<SubmissionDTO> {
+        const rows = await this.knex
+            .withSchema('public')
+            .select<DatabaseEntry[]>('id', 'updated', 'created_by', 'status', 'meta')
+            .from(this.TABLE_NAME)
+            .where({ id });
+
+        return this.entryToDTO(rows[0]);
```
Comment (norisalsaadie) [inline src/submission/repositories/xpub-submission-root.ts]: `return this.entryToDTO(rows[0]);`
Will this line throw an exception when entryToDTO tries to destructure if the select returns an empty array?

Comment (norisalsaadie) [reply src/submission/repositories/xpub-submission-root.ts]:
``` 
it('No results - DOES THIS BREAK???', async (): Promise<void> => {
            mockKnex.where = jest.fn().mockReturnValue([]);
            const repo = new XpubSubmissionRootRepository((mockKnex as unknown) as Knex);
            const result = await repo.findById(entryId);
            expect((result as SubmissionDTO).id).toEqual(databaseEntries[0].id);
});
```

Adding this test does break its during the destructure

**Comment by Cooryd** [comment]
CI is causing issues probably due to a hook issue yesterday. Will re-open pr

**Pull request closed without merging**